src/Projection/InMemoryProjectionManager.php

```diff
@@ -0,0 +1,195 @@
+    private const DEFAULT_SLEEP = 100000;
+    private const DEFAULT_PERSIST_BLOCK_SIZE = 1000;
+
+    /**
+     * @var EventStore
+     */
+    private $eventStore;
+
+    /**
+     * @var array
+     *
+     * key = projection name
+     * value = projection instance
+     */
+    private $projections = [];
+
+    public function __construct(EventStore $eventStore)
+    {
+        $this->eventStore = $eventStore;
+
+        while ($eventStore instanceof EventStoreDecorator) {
+            $eventStore = $eventStore->getInnerEventStore();
+        }
+
+        if (! $eventStore instanceof InMemoryEventStore) {
+            throw new Exception\InvalidArgumentException('Unknown event store instance given');
+        }
+    }
+
+    public function createQuery(): Query
+    {
+        return new InMemoryEventStoreQuery($this->eventStore);
+    }
+
+    public function createProjection(
+        string $name,
+        array $options = null
+    ): Projection {
+        $projection = new InMemoryEventStoreProjection(
+            $this->eventStore,
+            $name,
+            $options[self::OPTION_CACHE_SIZE] ?? self::DEFAULT_CACHE_SIZE,
+            $options[self::OPTION_SLEEP] ?? self::DEFAULT_SLEEP
+        );
+
+        if (! isset($this->projections[$name])) {
+            $this->projections[$name] = $projection;
+        }
+
+        return $projection;
+    }
+
+    public function createReadModelProjection(
+        string $name,
+        ReadModel $readModel,
+        array $options = null
+    ): ReadModelProjection {
+        $projection = new InMemoryEventStoreReadModelProjection(
+            $this->eventStore,
+            $name,
+            $readModel,
+            $options[self::OPTION_CACHE_SIZE] ?? self::DEFAULT_CACHE_SIZE,
+            $options[self::OPTION_PERSIST_BLOCK_SIZE] ?? self::DEFAULT_PERSIST_BLOCK_SIZE,
+            $options[self::OPTION_SLEEP] ?? self::DEFAULT_SLEEP
+        );
+
+        if (! isset($this->projections[$name])) {
+            $this->projections[$name] = $projection;
+        }
+
+        return $projection;
+    }
+
+    public function deleteProjection(string $name, bool $deleteEmittedEvents): void
+    {
+        throw new Exception\RuntimeException('Deleting a projection is not supported in ' . get_class($this));
+    }
+
+    public function resetProjection(string $name): void
+    {
+        throw new Exception\RuntimeException('Resetting a projection is not supported in ' . get_class($this));
+    }
+
+    public function stopProjection(string $name): void
+    {
+        throw new Exception\RuntimeException('Stopping a projection is not supported in ' . get_class($this));
+    }
+
+    public function fetchProjectionNames(?string $filter, bool $regex, int $limit, int $offset): array
+    {
+        if (null === $filter && $regex) {
+            throw new Exception\InvalidArgumentException('No regex pattern given');
+        }
+
+        if ($regex && false === @preg_match("/$filter/", '')) {
```

Comment:
If you use a single call with preg_grep with an try/catch by using set_error_handler, you can avoid that @ and a second regex parse process.